firebase-sessions/src/main/kotlin/com/google/firebase/sessions/SessionLifecycleClient.kt

```diff
@@ -0,0 +1,154 @@
+        }
+      }
+    }
+
+    fun handleSessionUpdate(sessionId: String) {
+      Log.i(TAG, "Session update received: $sessionId")
+      curSessionId = sessionId
+    }
+  }
+
+  /** The connection object to the [SessionLifecycleService]. */
+  private val serviceConnection =
+    object : ServiceConnection {
+      override fun onServiceConnected(className: ComponentName, serviceBinder: IBinder) {
+        Log.i(TAG, "Connected to SessionLifecycleService. Queue size ${queuedMessages.size}")
+        service = Messenger(serviceBinder)
+        serviceBound = true
+        val queueItr = queuedMessages.iterator()
+        for (msg in queueItr) {
+          Log.i(TAG, "sending queued message ${msg.what}")
+          sendLifecycleEvent(msg)
+          queueItr.remove()
+        }
+      }
+
+      override fun onServiceDisconnected(className: ComponentName) {
+        Log.i(TAG, "Disconnected from SessionLifecycleService")
+        service = null
+        serviceBound = false
+      }
+    }
+
+  /**
+   * Binds to the [SessionLifecycleService] and passes a callback [Messenger] that will be used to
+   * relay session updates to this client.
+   */
+  fun bindToService(appContext: Context): Unit {
+    Intent(appContext, SessionLifecycleService::class.java).also { intent ->
+      Log.i(TAG, "Binding service to application.")
+      // This is necessary for the onBind() to be called by each process
+      intent.setAction(android.os.Process.myPid().toString())
+      intent.putExtra(
+        SessionLifecycleService.CLIENT_CALLBACK_MESSENGER,
+        Messenger(ClientUpdateHandler())
+      )
+      appContext.bindService(
+        intent,
+        serviceConnection,
+        Context.BIND_IMPORTANT or Context.BIND_AUTO_CREATE
+      )
+    }
+  }
+
+  /**
+   * Should be called when any activity in this application process goes to the foreground. This
+   * will relay the event to the [SessionLifecycleService] where it can make the determination of
+   * whether or not this foregrounding event should result in a new session being generated.
+   */
+  fun foregrounded(activity: Activity): Unit {
+    sendLifecycleEvent(SessionLifecycleService.FOREGROUNDED)
+  }
+
+  /**
+   * Should be called when any activity in this application process goes from the foreground to the
+   * background. This will relay the event to the [SessionLifecycleService] where it will be used to
+   * determine when a new session should be generated.
+   */
+  fun backgrounded(activity: Activity): Unit {
+    sendLifecycleEvent(SessionLifecycleService.BACKGROUNDED)
+  }
+
+  /** Sends a message to the [SessionLifecycleService] with the given event code. */
+  private fun sendLifecycleEvent(messageCode: Int) {
+    sendLifecycleEvent(Message.obtain(null, messageCode, 0, 0))
+  }
+
+  /** Sends the given [Message] to the [SessionLifecycleService]. */
+  private fun sendLifecycleEvent(msg: Message) {
+    if (service != null) {
+      try {
+        service?.send(msg)
```

Comment:
If this succeeds, then why not drain the queue here as well?

Reply:
Updated to always try to drain the latest messages on send.